software/bsg_manycore_lib/bsg_tile_group_barrier.hpp

```diff
@@ -0,0 +1,260 @@
+class bsg_row_barrier {
+public:
+    volatile unsigned int     _local_alert = 0;
+    volatile unsigned char     _done_list[ BARRIER_X_DIM ] = {0};
+
+
+    bsg_row_barrier (){};
+
+
+    // Reinitializes the done_list to zero
+    void reset() {
+        for (int i = 0; i < BARRIER_X_DIM; i ++) {
+            unsigned char* done_list_ptr = const_cast<unsigned char*> (&_done_list[i]);
+            *done_list_ptr = 0;
+        }
+        return;
+    };
+
+
+    // send the sync signal to the center tile of the row
+    // executed by all tiles in the group.
+    void sync (unsigned char center_x_cord) {
+        //write to the corresponding done
+        volatile unsigned char *done_list_ptr = const_cast<unsigned char*> (reinterpret_cast<volatile unsigned char*> (bsg_remote_pointer(center_x_cord, bsg_y, &_done_list[bsg_x])));
+        *done_list_ptr = 1;
+        return;
+    };
+
+
+    // send alert to all of the tiles in the row 
+    void alert (){
+        for( int x = 0; x < BARRIER_X_DIM; x ++) {
+               volatile unsigned int *alert_ptr = const_cast<unsigned int*> (reinterpret_cast<volatile unsigned int*> (bsg_remote_pointer(x, bsg_y, &_local_alert)));
+               *alert_ptr = 1;
+        }
+        return;
+    };
+
+
+    // Poll the entire row barrier done_list until
+    // all tiles in row have sent their sync signal
+    // Executed by center tile in the row
+    void wait_on_sync() {
+        poll_range(this->_done_list, BARRIER_X_DIM);
+        return;
+    };
+
+
+    // wait on local alert to be set to the given value
+    void wait_on_alert (){
+        // wait until _local_alert flag is set to 1
+        bsg_wait_local_int( (int *) &(this->_local_alert), 1);
+        //re-initilized the flag to 0
+        unsigned int* alert_ptr = const_cast<unsigned int*> (&_local_alert);
+        *alert_ptr = 0;
+        return;
+    };
+};
+
+
+template <int BARRIER_Y_DIM>
+class bsg_col_barrier {
+public:
+    volatile unsigned int      _local_alert = 0;
+    volatile unsigned char      _done_list[ BARRIER_Y_DIM ] = {0};
+
+
+    bsg_col_barrier (){};
+
+
+    // Reinitializes the done_list to zero
+    void reset() {
+        for (int i = 0; i < BARRIER_Y_DIM; i ++) {
+            unsigned char* done_list_ptr = const_cast<unsigned char*> (&_done_list[i]);
+            *done_list_ptr = 0;
+        }
+        return;
+    };
+
+    // send the sync signal to the center tile of the column
+    // executed by all tiles in the center row
+    void sync(unsigned char center_x_cord, unsigned char center_y_cord ){
+        //write to the corresponding done
+        volatile unsigned char *done_list_ptr = const_cast<unsigned char*> (reinterpret_cast<volatile unsigned char*> (bsg_remote_pointer(center_x_cord, center_y_cord, &_done_list[bsg_y])));
+        *done_list_ptr = 1;
+
+        #ifdef BSG_BARRIER_DEBUG
+               //addr 0x0: row sync'ed
+               bsg_remote_ptr_io_store( IO_X_INDEX, 0x0, bsg_y);
+        #endif
```

Comment:
I would say the same about moving everything to C++ here, but I'm willing to let it slide because it's a debug statement.